### PR TITLE
Api integration

### PR DESCRIPTION
**Pull request by inetaa**
Legacy API integration, work in progress

### CODE REVIEW

File: app/src/main/java/lt/vilnius/tvarkau/API/ApiError.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package lt.vilnius.tvarkau.API;
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/ApiError.java]: Convention for package naming is lowercase, single word. If need to use 2 words, then just `firstsecond`

File: app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    private String session_id;
+    private String description;
+    private String type;
+    private String address;
+    private double x;
+    private double y;
+    private String[] photo;
+    private String email;
+    private String phone;
+    private String message_description;
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java]: Better use:

``` java
@SerializedName("message_description")
private String messageDescription;
```

File: app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+public class GetNewProblemParams {
+
+    private String session_id;
+    private String description;
+    private String type;
+    private String address;
+    private double x;
+    private double y;
+    private String[] photo;
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java:13]: Gson should be able to handle `List<String>`

File: app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    private String email;
+    private String phone;
+    private String message_description;
+
+    public GetNewProblemParams(String session_id, String description, String type, String address, double latitude,
+        double longitude, @Nullable String[] photo, @Nullable String email, @Nullable String phone,
+        @Nullable String messageDescription) {
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java:20]: When there are that many parameters, constructors become hardly usable. Better to use Builder pattern then.

Btw, IDE should be able to generate that, found this: https://www.jetbrains.com/help/idea/2016.2/replace-constructor-with-builder.html

File: app/src/main/java/lt/vilnius/tvarkau/API/GetProblemsParams.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+public class GetProblemsParams {
+
+    private int start;
+    private int limit;
+    private String description_filter;
+    private String type_filter;
+    private String address_filter;
+    private String reporter_filter;
+    private String date_filter;
+    private String status_filter;
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetProblemsParams.java]: Use `@SerializedName`. Also, there should be a naming policy for Gson to automatically map `type_filter <--> typeFilter`

File: app/src/main/java/lt/vilnius/tvarkau/API/GetProblemsParams.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+    public GetProblemsParams(int start, int limit, @Nullable String descriptionFilter, @Nullable String typeFilter,
+        @Nullable String addressFilter, @Nullable String reporterFilter, @Nullable String dateFilter,
+        @Nullable String statusFilter) {
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetProblemsParams.java:18]: Again, Builder would be nice

File: app/src/main/java/lt/vilnius/tvarkau/API/LegacyApiModule.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        return new OkHttpClient.Builder()
+            .authenticator(new TokenAuthenticator())
+            .addNetworkInterceptor(new Interceptor() {
+                @Override public Response intercept(Chain chain) throws IOException {
+                    Request.Builder requestBuilder = chain.request().newBuilder();
+                    requestBuilder.header("Content-Type", "application/json");
+                    return chain.proceed(requestBuilder.build());
+                }
+            })
+            .addNetworkInterceptor(new HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY))
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/LegacyApiModule.java]: Logging should be disabled in production. There's flag `BuildConfig.DEBUG` - could turn on logging only when it is `true`

File: app/src/main/java/lt/vilnius/tvarkau/API/LegacyApiModule.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+            .addInterceptor(new TokenInterceptor())
+            .addNetworkInterceptor(new StethoInterceptor())
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/LegacyApiModule.java]: Same here, should be removed for production

File: app/src/main/java/lt/vilnius/tvarkau/API/LegacyApiModule.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+            .build();
+    }
+
+    @Provides
+    @Singleton
+    public Retrofit provideRetrofit(OkHttpClient okHttpClient) {
+
+        Gson gson = new GsonBuilder()
+            .registerTypeAdapter(LocalDateTime.class, new LocalDateTimeSerializer())
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/API/LegacyApiModule.java:70]: Adding this should automatically map JSON `type_filter` to Java `typeFilter`:

```
Gson gson = new GsonBuilder()
            .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
```

File: app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java [hunk 10]
```diff
@@ -149,14 +198,86 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
+
+    private ProgressDialog createProgressDialog() {
+        ProgressDialog progressDialog = new ProgressDialog(this);
+        progressDialog.setMessage(getString(R.string.sending_problem));
+        progressDialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
+        return progressDialog;
+    }
+
+    private Boolean validateProblemInputs() {
+        Boolean addressIsValid = false;
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java]: Use primitive types (`boolean`) unless you really need boxed ones

File: app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java [hunk 11]
```diff
@@ -231,8 +350,22 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+                            Bitmap bitmap = MediaStore.Images.Media.getBitmap(this.getContentResolver(), uriPath);
+                            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+                            bitmap.compress(Bitmap.CompressFormat.JPEG, 60, byteArrayOutputStream);
+                            byte[] byteArrayImage = byteArrayOutputStream.toByteArray();
+                            String encodedImage = Base64.encodeToString(byteArrayImage, Base64.NO_WRAP);
+                            photos[i] = encodedImage;
+                        } catch (IOException e) {
+                            e.printStackTrace();
+                        }
+                    }
```
Comment (mkuprionis) [inline app/src/main/java/lt/vilnius/tvarkau/NewProblemActivity.java]: This should happen on background thread - might take time and UI might freeze

File: app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java (hunk 2 above)
Comment (ViliusKraujutis) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java]: Also, there's `setFieldNamingPolicy`, which might be convenient http://stackoverflow.com/a/11810183/517381

File: app/src/main/java/lt/vilnius/tvarkau/API/GetProblemsParams.java (hunk 5 above)
Comment (ViliusKraujutis) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetProblemsParams.java]: :+1:

File: app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java (hunk 2 above)
Comment (ViliusKraujutis) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java]: oh, I see, Mindaugas mentioned that below. :+1:

File: app/src/main/java/lt/vilnius/tvarkau/fragments/ProblemsListFragment.java [hunk 12]
```diff
@@ -1,45 +1,250 @@
...
+    @Subscribe
+    public void onNewProblemAddedEvent(NewProblemAddedEvent event) {
+        getData(0);
+    }
+
     @Override
-    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
-        super.onActivityCreated(savedInstanceState);
+    public void onResume() {
+        getData(0);
```
Comment (ViliusKraujutis) [inline app/src/main/java/lt/vilnius/tvarkau/fragments/ProblemsListFragment.java:228]: Shouldn't it clear and reload 1st page data here?

Comment (inetaa) [inline app/src/main/java/lt/vilnius/tvarkau/fragments/ProblemsListFragment.java:228]: Yes, you are right, it's fixed on another branch.

**Comment by ViliusKraujutis** [comment]
:+1:

File: app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java (hunk 3 above)
Comment (inetaa) [inline app/src/main/java/lt/vilnius/tvarkau/API/GetNewProblemParams.java:13]: Kept String[] for now as we are using String arrays to set photos in UI

File: app/src/main/java/lt/vilnius/tvarkau/API/LegacyApiModule.java (hunk 9 above)
Comment (inetaa) [inline app/src/main/java/lt/vilnius/tvarkau/API/LegacyApiModule.java:70]: Fixed

**Pull request merged**